lib/api/objectGetRetention.js

```diff
@@ -0,0 +1,104 @@
+const async = require('async');
+const { errors, s3middleware } = require('arsenal');
+
+const { decodeVersionId, getVersionIdResHeader }
+    = require('./apiUtils/object/versioning');
+
+const { metadataValidateBucketAndObj } = require('../metadata/metadataUtils');
+const { pushMetric } = require('../utapi/utilities');
+const collectCorsHeaders = require('../utilities/collectCorsHeaders');
+
+const { convertToXml } = s3middleware.retention;
+
+/**
+ * Object Get Retention - Return retention info for object
+ * @param {AuthInfo} authInfo - Instance of AuthInfo class with requester's info
+ * @param {object} request - http request object
+ * @param {object} log - Werelogs logger
+ * @param {function} callback - callback to server
+ * @return {undefined}
+ */
+function objectGetRetention(authInfo, request, log, callback) {
+    log.debug('processing request', { method: 'objectGetRetention' });
+
+    const { bucketName, objectKey } = request;
+
+    const decodedVidResult = decodeVersionId(request.query);
+    if (decodedVidResult instanceof Error) {
+        log.trace('invalid versionId query', {
+            versionId: request.query.versionId,
+            error: decodedVidResult,
+        });
+        return process.nextTick(() => callback(decodedVidResult));
+    }
+    const reqVersionId = decodedVidResult;
+
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        objectKey,
+        requestType: 'objectGetRetention',
+        versionId: reqVersionId,
+    };
+
+    return async.waterfall([
+        next => metadataValidateBucketAndObj(metadataValParams, log,
+            (err, bucket, objectMD) => {
+                if (err) {
+                    log.trace('request authorization failed',
+                        { method: 'objectGetRetention', error: err });
+                    return next(err);
+                }
+                if (!objectMD) {
+                    const err = reqVersionId ? errors.NoSuchVersion :
+                        errors.NoSuchKey;
+                    log.trace('error no object metadata found',
+                        { method: 'objectGetRetention', error: err });
+                    return next(err, bucket);
+                }
+                if (objectMD.isDeleteMarker) {
+                    if (reqVersionId) {
+                        log.trace('requested version is delete marker',
+                            { method: 'objectGetRetention' });
+                        return next(errors.MethodNotAllowed);
+                    }
+                    log.trace('most recent version is delete marker',
+                        { method: 'objectGetRetention' });
+                    return next(errors.NoSuchKey);
```

Comment:
As far as I can tell reading the tests and this code, this is all untested.

Reply:
Tests are in a followup PR: https://github.com/scality/cloudserver/pull/2609